### PR TITLE
Reduces redundancy by using only one function to handle all outgoing HTML mail

### LINKED ISSUE #26: HTML handler should respect other [email, system] handlers
The ```html_email_handler_email_hook()``` is ignoring parameters added/modified by other handlers registered for the ```[email, system]``` hook.

E.g. this modified value gets ignored: https://github.com/Elgg/Elgg/blob/1.9/engine/lib/comments.php#L224

The values should be extracted from the third ```$return``` parameter instead of the fourth ```$params``` parameter, which contains the unmodified values.

### LINKED ISSUE #27: Asynchronous notifications should trigger [email, system] hook
Problem
-----------
Sending of asynchronous notifications through the Elgg 1.9 notifications service.

**Default core workflow is:**
 1. Notifications service triggers ```[send, notification:email]``` hook
 2. The hook invokes ```elgg_send_email()```
 3. ```elgg_send_email()``` handler triggers ```[email, system]``` hook
 4. Handlers now have the ability to modify email before sending
 5. Emails are sent

**html_email_handler workflow is:**
 1. Notifications service triggers ```[send, notification:email]``` hook
 2. The hook invokes ```html_email_handler_send_email_notifications_hook()```
 3. ```html_email_handler_send_email_notifications_hook()``` calls ```html_email_handler_send_email()```
 4. - - -
 5. Emails are sent, but ```[email, system]``` hook was never triggered!

Solution
-----------

At first I was thinking that this is easy to fix by triggering ```[email, system]``` *also* from ```html_email_handler_send_email()```. That handler is however itself registered for the ```[email, system]``` for the instant notifications (synchronously sent). So this approach would cause an infinite loop.

Possible solution could be to use the default core handler and do the stuff provided by  ```html_email_handler_send_email_notifications_hook``` in the ```[email, system]``` hook instead of ```[send, notification:email]``` hook.

The only difference then would be that it would get called both for synchronously and asynchronously sent emails. Currently it gets called only for the latter. Is there a reason for this?

### PR DESCRIPTION
**Pull request by juho-jaakkola**
Also adds an attachment to the test.

Fixes #26
Fixes #27

### CODE REVIEW

**Comment by juho-jaakkola** [comment]
I think this is good to go!

**Comment by jeabakker** [comment]
the way you reused the function name html_email_handler_send_email() to be a plugin hook handler will break every plugin that uses that function (digest, newsletter to name a few).

this is bc-breaking ;) we don't want to do that.

Also why have an extra file for the attachment test, why not send for example the manifest

File: lib/hooks.php [hunk 1]
```diff
@@ -4,118 +4,25 @@
...
-		"plaintext_message" => $message->body
-	);
-	
-	if (!empty($params) && is_array($params)) {
-		$options = array_merge($options, $params);
-	}
-	
-	return html_email_handler_send_email($options);
+function html_email_handler_email_hook($hook, $type, $options, $params) {
+	return html_email_handler_send_email($return);
```
Comment (jeabakker) [inline lib/hooks.php]: $return is undefined should be $options

File: lib/functions.php [hunk 2]
```diff
@@ -157,9 +156,23 @@ function html_email_handler_send_email(array $options = null) {
...
 	// Use attachments headers for real only if they are valid
 	if (!empty($attachments)) {
-		$headers .= "Content-Type: multipart/mixed; boundary=\"mixed--" . $boundary . "\"" . PHP_EOL . PHP_EOL;
+		$headers['Content-Type'] = "multipart/mixed; boundary=\"mixed--{$boundary}\"";
 	} else {
-		$headers .= "Content-Type: multipart/alternative; boundary=\"" . $boundary . "\"" . PHP_EOL . PHP_EOL;
+		$headers['Content-Type'] = "multipart/alternative; boundary=\"{$boundary}\"";
+	}
+	
+	global $CONFIG;
```
Comment (jeabakker) [inline lib/functions.php]: please don't use $CONFIG but use elgg_get_config() instead

File: lib/functions.php [hunk 3]
```diff
@@ -80,6 +77,9 @@ function html_email_handler_send_email(array $options = null) {
 		$options["bcc"] = array($options["bcc"]);
 	}
 	
+	$options['html_message'] = html_email_handler_make_html_body($options);
+	$options['plaintext_message'] = $options['body'];
```
Comment (juho-jaakkola) [inline lib/functions.php]: I guess these two should be added only if both of them are empty? Should I also check whether the global HTML email setting has been enabled?

Comment (jeabakker) [inline lib/functions.php]: i think they should only be filled when not empty.

no need to check the plugin setting, this function is there to make sure you can send html mail regardless of the setting. Only the hooks need to respect the plugin setting

**Comment by jeabakker** [comment]
please check the one tiny issue in scrutinizer, if fixed i can merge

**Comment by juho-jaakkola** [comment]
done

**Comment by jeabakker** [comment]
@Facyla could you maybe have a look at this pr. The attachements still work according to juho but since you build it you can see if it all works

**Comment by Facyla** [comment]
Hi,

I'm sorry i will not have time today (and probably this whole week) to test it in detail, so a quick note after reading the changes : 

In hook, the "attachments" now get passed through $options, and not $params : as far as i read the discussion that was precisely what has been improved through this commit, but it modifies a little the way attachments are passed.
A further test would be checking that attachments also gets passed when using : notify_user() function with $params['attachments']
=> If this is OK, it's fine for me.

Sorry for not testing it myself...
Thank you,
Florian

**Comment by jeabakker** [comment]
@Facyla no problem, thanks for the tip. 

I think you have a point maybe @juho-jaakkola can have a look at it

**Comment by juho-jaakkola** [comment]
The ```$params``` should be used only for original core parameters (since 2012: 16cf5867f33bada9837158272a8e944347a4713b).

I suggest using it for now but giving a deprecation warning if the ```'attachments'``` option is found from there.

**Comment by Facyla** [comment]
Ok for deprecation warning, that was mainly for bc reasons, but i totally agree that some behaviours require refactoring :)

**Comment by juho-jaakkola** [comment]
> the "attachments" now get passed through $options, and not $params

This doesn't matter because Elgg core populates them both with the same values (e.g. the parameters passed into ```notify_user()```).

The ```notify_user()``` function does however pass the attachments under ```$options['params']```, so I added a fix that looks for them from ```$options['params']['attachments']``` instead of ```$options['attachments']```, so now they also work.

I also added a test that sends an email through ```notify_user()``` with an attachment.

**Pull request merged**